### PR TITLE
Chapter 11: Account activation

### PR DESCRIPTION
**Pull request by baaohuii**

### CODE REVIEW

**Comment by baaohuii** [comment]
ready

File: app/assets/stylesheets/account_activations.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the AccountActivations controller here.
```
Comment (manhnt-0960) [inline app/assets/stylesheets/account_activations.scss]: let's delete this empty file

File: app/controllers/account_activations_controller.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    　user.activate　　
+      log_in user
+      flash[:success] = "Account activated!"
+      redirect_to user
+    else
+      flash[:danger] = "Invalid activation link"
+      redirect_to root_url
+    end
+  end
+end
```
Comment (manhnt-0960) [inline app/controllers/account_activations_controller.rb]: add new line at end of file

File: app/helpers/account_activations_helper.rb [hunk 3]
```diff
@@ -0,0 +1,2 @@
+module AccountActivationsHelper
```
Comment (manhnt-0960) [inline app/helpers/account_activations_helper.rb]: please delete this empty file

File: app/models/user.rb [hunk 4]
```diff
@@ -20,7 +23,12 @@ class User < ApplicationRecord
   private  
 
   def downcase_email
-    email.downcase! 
+    self.email = email.downcase
```
Comment (manhnt-0960) [inline app/models/user.rb]:
```suggestion
    email.downcase!
```

File: app/models/user.rb [hunk 5]
```diff
@@ -38,15 +46,24 @@ def new_token
...
-    update_attribute(:remember_digest, User.digest(remember_token))
+    update_attribute :remember_digest, User.digest(remember_token)
   end
 
-  def authenticated?(remember_token)
-    return false if remember_digest.nil?
-    BCrypt::Password.new(remember_digest).is_password?(remember_token)
+  def authenticated?(attribute, token)
+    digest = send("#{attribute}_digest")
+    return false if digest.nil?
```
Comment (manhnt-0960) [inline app/models/user.rb]:
```suggestion
    return false unless digest
```

File: config/routes.rb [hunk 6]
```diff
@@ -9,4 +9,5 @@
   delete "/logout", to: "sessions#destroy"
   get "users/show"
   resources :users
+  resources :account_activations, only: [:edit]
```
Comment (manhnt-0960) [inline config/routes.rb]:
```suggestion
  resources :account_activations, only: :edit
```

**Comment by baaohuii** [comment]
ready

File: app/controllers/account_activations_controller.rb [hunk 7]
```diff
@@ -0,0 +1,14 @@
+class AccountActivationsController < ApplicationController
+  def edit
+  user = User.find_by email: params[:email]
```
Comment (saobv-2223) [inline app/controllers/account_activations_controller.rb]:
```suggestion
    user = User.find_by email: params[:email]
```

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -5,11 +5,12 @@ class UsersController < ApplicationController
   before_action :admin_user, only: :destroy
 
   def index
-    @users = User.sort_by_name.page params[:page]
+    @users = User.where activated: FILL_IN .paginate page: params[:page]
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: please, using kaminari gem to paginate

File: app/mailers/user_mailer.rb [hunk 9]
```diff
@@ -0,0 +1,13 @@
+class UserMailer < ApplicationMailer
+
+  def account_activation(user)
```
Comment (saobv-2223) [inline app/mailers/user_mailer.rb]:
```suggestion
  def account_activation(user)
```

File: app/models/user.rb [hunk 10]
```diff
@@ -38,15 +46,24 @@ def new_token
...
+    update_attribute :remember_digest, User.digest(remember_token)
   end
 
-  def authenticated?(remember_token)
-    return false if remember_digest.nil?
-    BCrypt::Password.new(remember_digest).is_password?(remember_token)
+  def authenticated?(attribute, token)
+    digest = send("#{attribute}_digest")
+    return false unless digest
+    BCrypt::Password.new(digest).is_password?(token)
```
Comment (saobv-2223) [inline app/models/user.rb]:
```suggestion
    return false unless digest
    
    BCrypt::Password.new(digest).is_password?(token)
```

File: app/views/user_mailer/account_activation.html.erb [hunk 11]
```diff
@@ -0,0 +1,9 @@
+<h1>Sample App</h1>
+<p>Hi <%= @user.name %>,</p>
+
+<p>
+Welcome to the Sample App! Click on the link below to activate your account:
+</p>
```
Comment (saobv-2223) [inline app/views/user_mailer/account_activation.html.erb]:
```suggestion
<p>
  Welcome to the Sample App! Click on the link below to activate your account:
</p>
```

**Comment by baaohuii** [comment]
ready

**Comment by manhnt-0960** [comment]
merged

**Pull request merged**